// approuter/lib/configuration/env-config.js
'use strict';

var _ = require('lodash');
var logger = require('../utils/logger').getLogger('/Configuration');
var path = require('path');
var validators = require('./validators');
var VError = require('verror').VError;
var xsenv = require('sap-xsenv');
var configurationUtils = require('../utils/configuration-utils');

module.exports = {
  load: loadAll
};

function loadAll(workingDir) {
  var filePath = path.join(workingDir, 'default-env.json');
  xsenv.loadEnv(filePath);

  var envConfig = {};
  envConfig.destinations = loadDestinations();
  envConfig.additionalHeaders = loadAdditionalHeaders();
  envConfig.plugins = loadPlugins(envConfig.destinations);
  envConfig.clickjackCheckWhitelist = loadClickjackWhitelist();
  envConfig.wsAllowedOrigins = loadWebSocketWhitelist();

  var grouped = loadGroupedConfigurations();
  envConfig = _.merge(envConfig, grouped);

  return envConfig;
}

function loadDestinations() {
  var destinations = loadJsonVar('destinations');
  if (!destinations) {
    logger.info('Using empty destinations to run');
    return {};
  }
  validators.validateEnvDestinations(destinations);
  applyDestinationDefaults(destinations);
  return _.indexBy(destinations, 'name');
}

function loadAdditionalHeaders() {
  var additionalHeaders = loadJsonVar('httpHeaders');
  if (!additionalHeaders) {
    return [];
  }
  validators.validateHeaders(additionalHeaders);
  return additionalHeaders;
}

function loadPlugins(destinations) {
  var plugins = loadJsonVar('plugins');
  if (!plugins) {
    return null;
  }
  validators.validatePlugins(plugins, destinations);
  plugins.forEach(function (currentPlugin) {
    currentPlugin.source = configurationUtils.constructRegExp(currentPlugin.source);
  });
  return plugins;
}

function loadClickjackWhitelist() {
  return loadOriginsWhitelist('CJ_PROTECT_WHITELIST');
}

function loadWebSocketWhitelist() {
  return loadOriginsWhitelist('WS_ALLOWED_ORIGINS');
}

function loadOriginsWhitelist(envVar) {
  var whitelist = loadJsonVar(envVar);
  if (!whitelist) {
    return null;
  }
  validators.validateWhitelist(whitelist);
  whitelist.forEach(function (item) {
    if (typeof item.port === 'string') {
      item.port = parseInt(item.port);
    }
  });
  return whitelist;
}

function loadGroupedConfigurations() {
  var grouped = {
    sessionTimeout: loadJsonVar('SESSION_TIMEOUT'),
    incomingConnectionTimeout: loadJsonVar('INCOMING_CONNECTION_TIMEOUT'),
    jwtRefresh: loadJsonVar('JWT_REFRESH')
  };

  grouped = _.pick(grouped, function (value) {
    return (value || value === '' || value === 0);
  });

  if (!grouped.hasOwnProperty('jwtRefresh')) {
    grouped.jwtRefresh = 5; // 5 is the default JWT_REFRESH
  }
  var sendXFrameOptions = loadJsonVar('SEND_XFRAMEOPTIONS');
  grouped.sendXFrameOptions = (typeof sendXFrameOptions !== 'undefined') ? sendXFrameOptions : true;
  validators.validateEnvironmentSettings(grouped);

  return grouped;
}

function loadJsonVar(envVar) {
  if (envVar in process.env) {
    try {
      return JSON.parse(process.env[envVar]);
    } catch (e) {
      throw new VError(e, 'Invalid value for environment variable %s', envVar);
    }
  }
}

function applyDestinationDefaults(destinations) {
  destinations.forEach(function (dest) {
    _.defaults(dest, {
      timeout: 30000
    });
  });
}
