// approuter/lib/utils/uaa-utils.js
'use strict';

var _ = require('lodash');
var xsenv = require('sap-xsenv');
var path = require('path');
var logger = require('./logger').getLogger('/Configuration');


module.exports = {
  getOptions: function (workingDir) {
    var pathToDefaultServices = path.join(workingDir, 'default-services.json');
    var uaa;
    try {
      uaa = xsenv.getServices({ uaa: matchesUaa }, pathToDefaultServices).uaa;
    } catch (e) {
      logger.error(e, 'Cannot find service uaa in environment');
      return {};
    }

    return {
      authorizationURL: _.trimRight(uaa.url, '/') + '/oauth/authorize',
      tokenURL: _.trimRight(uaa.url, '/') + '/oauth/token',
      logoutURL: _.trimRight(uaa.url, '/') + '/logout.do',
      clientID: uaa.clientid,
      clientSecret: uaa.clientsecret,
      xsappname: uaa.xsappname
    };
  }
};

function matchesUaa(service) {
  var uaaName = process.env.UAA_SERVICE_NAME;
  if (uaaName) {
    return service.name === uaaName;
  }

  if (service.tags && _.contains(service.tags, 'xsuaa')) {
    return true;
  }
  if (service.label === 'user-provided' && service.credentials.tags && _.contains(service.credentials.tags, 'xsuaa')) {
    return true;
  }
  return false;
}
