// sap-site-entry/lib/config.js
'use strict';

const services = JSON.parse(process.env.VCAP_SERVICES);
var service;

var isArrayInclude = function(arr, item) {
    var isFound = false;
    if (arr && item) {
        for (var index in arr) {
            if (arr[index] === item) {
                return true;
            }
        }
    }
    return isFound;
}

for (var property in services) {
    if (services[property] && services[property][0] && services[property][0].tags) {
        var tags = services[property][0].tags;
        if (isArrayInclude(tags, "portal")) {
            service = services[property][0];
        }
    }
}

if (!service) {
    console.error("can't bind to persistency service.!");
    throw Error("can't startup");
}

class Config {

    static getInstanceId() {
        return service['credentials']["instanceId"];
    }

    static getBindingId() {
        return service['credentials']["bindingId"];
    }
    static getUrl() {
        return service['credentials']["url"];
    }

    static getAuthUser() {
        return service['credentials']["user"];
    }

    static getAuthPassword() {
        return service['credentials']["password"];
    }

}

module.exports = Config;
