// approuter/lib/passport/utils.js
'use strict';

var headerUtil = require('../utils/header-util');
var urlUtils = require('../utils/url-utils');
var jwtDecode = require('jwt-decode');

module.exports = {
  secureCookie: function (req) {
    var secure;
    switch (process.env.SECURE_SESSION_COOKIE) {
    case 'true':
      secure = true;
      break;
    case 'false':
      secure = false;
      break;
    default:
      secure = req.protocol !== 'http';
    }
    req.session.cookie.secure = secure;
  },

  storeToken: function(req, accessToken, refreshToken, params, profile, done) {
    req.session.regenerate(function(err) {
      module.exports.secureCookie(req);
      var options = { accessToken: accessToken, refreshToken: refreshToken, expiresIn: params.expires_in, scope: params.scope };

      req.session.user = module.exports.getUserProperties(options);
      req.session.save();

      var user = { name: 'approuter does not require that, but passport does' };
      done(err, user);
    });
  },

  getUserProperties: function(options) {
    return {
      name: jwtDecode(options.accessToken).user_name,
      token: {
        accessToken: options.accessToken,
        expiryDate: module.exports.getExpiresAt(options.expiresIn).getTime(),
        refreshToken: options.refreshToken
      },
      scopes: options.scope ? options.scope.trim().split(' ') : []
    };
  },

  getExpiresAt: function(duration) {
    // respect network latency if possible
    var expiresIn;
    if (duration > 180) {
      expiresIn = duration - 120;
    } else {
      expiresIn = duration;
    }
    return new Date(new Date().getTime() + expiresIn * 1000);
  },

  getCallBackUrl: function(req) {
    var appRouterUrl = urlUtils.buildAppRouterUrl(req);
    var loginCallback = req.app.get('routerConfig').appConfig.login.callbackEndpoint;
    return urlUtils.join(appRouterUrl, loginCallback);
  },

  loadOauthOptions: function(req) {
    var options = req.app.get('routerConfig').uaaOptions;

    options.callbackURL = this.getCallBackUrl(req);
    options.passReqToCallback = 'true';
    options.customHeaders = headerUtil.getBasicAuthHeader(options.clientID, options.clientSecret);
    headerUtil.updateSapPassport(req.headers, options.customHeaders);
    return options;
  }
};
