// approuter/lib/bootstrap.js
'use strict';
process.env.NO_DEPRECATION = 'express-session';

var https = require('https');
var expressSession = require('express-session');
var passport = require('passport');
var xsenv = require('sap-xsenv');

var connect = require('./connect/connect');
var agents = require('./backend-request/agents');
var additionalHeadersMiddleware = require('./middleware/additional-headers-middleware');
var authorizationMiddleware = require('./middleware/authorization-middleware');
var compressionMiddleware = require('./middleware/compression-middleware');
var configuration = require('./configuration');
var cookieUtils = require('./utils/cookie-utils');
var errorHandler = require('./middleware/error-handler');
var loginMiddleware = require('./middleware/login-middleware');
var logging = require('./utils/logger');
var logoutMiddleware = require('./middleware/logout-middleware');
var memstoreConfig = require('./configuration/memstore-config');
var pathRewritingMiddleware = require('./middleware/path-rewriting-middleware');
var loginCallbackMiddleware = require('./middleware/login-callback-middleware');
var pluginEndpointMiddleware = require('./middleware/plugin-middleware');
var jwtRefreshMiddleware = require('./middleware/jwt-refresh-middleware');
var requestHandler = require('./middleware/request-handler');
var staticResourceHandler = require('./middleware/static-resource-handler');
var secureCookieMiddleware = require('./middleware/secure-cookie-middleware');
var sessionCookieMiddleware = require('./middleware/session-cookie-middleware');
var traceRequestMiddleware = require('./middleware/trace-request-middleware');
var welcomePageMiddleware = require('./middleware/welcome-page-middleware');
var xsrfTokenMiddleware = require('./middleware/xsrf-token-middleware');
var whitelistServiceMiddleware = require('./middleware/whitelist-service-middleware');
var extensions = require('./extensions/extensions');

module.exports = function bootstrap(options) {
  var certificates = xsenv.loadCertificates();
  agents.httpsAgent.options.ca = certificates;
  https.globalAgent.options.ca = certificates;

  var routerConfig = configuration(options);
  var ext = extensions(options.extensions);

  var app = connect();
  ext.insertMiddleware('first', app);

  var sessionSecret = cookieUtils.generateSessionSecret();

  app.use(logging.getExpressMiddleware());
  app.use(traceRequestMiddleware);
  app.use(passport.initialize());
  app.use(passport.session());

  var memoryStore = memstoreConfig.getMemoryStore(routerConfig);
  var cookieName = cookieUtils.getSessionCookieName();

  app.set('cookieName', cookieName);
  app.set('memoryStore', memoryStore);
  app.set('routerConfig', routerConfig);
  app.set('sessionCookieKey', sessionSecret);
  app.set('trust proxy', true);

  app.use(welcomePageMiddleware);
  app.use(pathRewritingMiddleware);
  app.use(expressSession({
    name: cookieName,
    resave: false,
    saveUninitialized: false,
    secret: sessionSecret,
    store: memoryStore
  }));
  app.use(secureCookieMiddleware);
  app.use(additionalHeadersMiddleware);

  app.use(routerConfig.appConfig.login.callbackEndpoint, loginCallbackMiddleware);
  app.use(loginMiddleware);
  if (routerConfig.jwtRefresh !== 0) {
    app.use(jwtRefreshMiddleware);
  }
  app.use(authorizationMiddleware);
  app.use(xsrfTokenMiddleware);

  var metadataEndpointOptions = routerConfig.appConfig.pluginMetadataEndpoint;
  metadataEndpointOptions && app.use(metadataEndpointOptions, pluginEndpointMiddleware);

  var logout = routerConfig.appConfig.logout;
  logout && logout.logoutEndpoint && app.use(logout.logoutEndpoint, logoutMiddleware);

  var whitelistService = routerConfig.appConfig.whitelistService;
  whitelistService && app.use(whitelistService.endpoint, whitelistServiceMiddleware);

  var compressionOptions = routerConfig.appConfig.compression;
  compressionOptions.enabled && app.use(compressionMiddleware(compressionOptions));

  ext.insertMiddleware('beforeRequestHandler', app);
  app.use(staticResourceHandler);
  app.use(sessionCookieMiddleware(cookieName));

  app.use(requestHandler);

  ext.insertMiddleware('beforeErrorHandler', app);
  app.use(errorHandler);

  return app;
};
