// sap-xsssecure/index.js
'use strict';


function h(i, l) {
  var g = i.toString(16);
  if (l) {
    while (l > g.length) {
      g = '0' + g;
    }
  }
  return g;
}
var r = /[\x00-\x2b\x2f\x3a-\x40\x5b-\x5e\x60\x7b-\xff\u2028\u2029]/g; // eslint-disable-line no-control-regex
var a = /[\x00-\x08\x0b\x0c\x0e-\x1f\x7f-\x9f]/; // eslint-disable-line no-control-regex
var H = {
  '<': '&lt;',
  '>': '&gt;',
  '&': '&amp;',
  '"': '&quot;'
};
var f = function(g) {
  var E = H[g];
  if (!E) {
    if (a.test(g)) {
      E = '&#xfffd;';
    } else {
      E = '&#x' + h(g.charCodeAt(0)) + ';';
    }
    H[g] = E;
  }
  return E;
};

function encodeHTML(S) {
  return S.replace(r, f);
}

var b = /[\x00-\x2b\x2d\x2f\x3a-\x40\x5b-\x5e\x60\x7b-\xff\u2028\u2029]/g; // eslint-disable-line no-control-regex
var j = {};
var J = function(g) {
  var E = j[g];
  if (!E) {
    var i = g.charCodeAt(0);
    if (i < 256) {
      E = '\\x' + h(i, 2);
    } else {
      E = '\\u' + h(i, 4);
    }
    j[g] = E;
  }
  return E;
};

function encodeJS(S) {
  return S.replace(b, J);
}

var c = /[\x00-\x2c\x2f\x3a-\x40\x5b-\x5e\x60\x7b-\uffff]/g; // eslint-disable-line no-control-regex
var u = {};
var U = function(g) {
  var E = u[g];
  if (!E) {
    var i = g.charCodeAt(0);
    if (i < 128) {
      E = '%' + h(i, 2);
    } else if (i < 2048) {
      E = '%' + h((i >> 6) | 192, 2) + '%' + h((i & 63) | 128, 2);
    } else {
      E = '%' + h((i >> 12) | 224, 2) + '%' + h(((i >> 6) & 63) | 128, 2) + '%' + h((i & 63) | 128, 2);
    }
    u[g] = E;
  }
  return E;
};

function encodeURL(S) {
  return S.replace(c, U);
}

var d = /[\x00-\x2f\x3a-\x40\x5b-\x60\x7b-\xff\u2028\u2029][0-9A-Fa-f]?/g; // eslint-disable-line no-control-regex
var C = function(g) {
  var i = g.charCodeAt(0);
  if (g.length === 1) {
    return '\\' + h(i);
  } else {
    return '\\' + h(i) + ' ' + g.substr(1);
  }
};

function encodeCSS(S) {
  return S.replace(d, C);
}


module.exports.encodeHTML = encodeHTML;
module.exports.encodeXML = encodeHTML;
module.exports.encodeJS = encodeJS;
module.exports.encodeURL = encodeURL;
module.exports.encodeCSS = encodeCSS;
